FIX: [Web App] Display different 404 component/page depending on the app the user is in

samuko-things opened the pull request:
**This pull request makes the following changes**
* Fixes code-collabo/collabo-community-app#58 

**General checklist**
- [x] File or folder now contains changes as specified in the issue i worked on
- [x] I have linked the issue I worked on to this pull request submitted by me

**Testing checklist**
- [x] Dummy content provided in the app is what is used, no figma design added
- [x] 3 different 404 pages show up for the different apps within our 3-in-1 app
- [x] Nextjs layout and pages structure approach is used
- [x] 404 page for community routes uses the page layout for community routes. And as well display dummy 404 content/component for community.
- [x] 404 page for code-collabo routes uses the page layout for code-collabo routes. And as well display dummy 404 content/component for code-collabo.
- [x] 404 page for other-subcommunity routes uses the page layout for other-subcommunity routes. And as well display dummy 404 content/component for other-subcommunity.
- [x] I certify that I ran my checklist

Ping @code-collabo/community-mentors

samuko-things committed:
added 404 dummy page for code-collabo route

samuko-things commented:
@Ifycode, @code-collabo/community-mentors, I have added the dummy 404 page for the code-collabo routes. pls do check it out

Ifycode commented:
Ok, this is a nice first step @samuko-things 
And code looks clean.
Have you tested it for code-collabo routes? And do it work on the page?

samuko-things committed:
added 404 dummy page for community route

samuko-things commented:
> Ok, this is a nice first step @samuko-things And code looks clean. Have you tested it for code-collabo routes? And do it work on the page?



> Ok, this is a nice first step @samuko-things And code looks clean. Have you tested it for code-collabo routes? And do it work on the page?

I've tested it with the .../code-collabo/404 route and it works and displays the page.

samuko-things commented:
@Ifycode, @code-collabo/community-mentors, I have added the dummy 404 page for the community routes. pls do check it out

samuko-things committed:
updated 404 dummy page name for community route

samuko-things committed:
updated 404 dummy page name for community route

samuko-things commented:
@Ifycode, https://github.com/orgs/code-collabo/teams/community-mentors, I have added the dummy 404 page for the other-subcommunity routes. pls do check it out. 

I also updated the 404 component name of the community route.

> pls do not mind the commit name, it was a mistake on my end

Ifycode reviewed (changes requested):
**_Review comment 1 of 2_**

So this is a great start @samuko-things but **_the 404 pages don't work yet_**. **_How to test what I mean?_** - Before you created this 3 404.tsx files, there is a 404.tsx that is directly in the root of the pages folder. Remove that particular 404.tsx file and then try to access a page that does not exist in the browser. You will see that nextjs default way of showing 404 takes over. 

That is to tell you that what is still showing up before was only that 404.tsx file that you deleted from the root of the pages folder.

Another confirmation is to restore the deleted 404.tsx file that is in the root of the pages folder. Check a non-exisiting route, you will see that only one text always shows up for 404: "Collabo Community Main App 404 page content". If you **_check the different 404 components you imported, you will see that different content is supposed to appear_** for the different types of 404.

#
**_Review comment 2 of 2_**

Also **_the difference in layout is supposed to reflect in the 404 pages too_** based on the part of the app that they take care of, then **_the specific dummy 404 page content should show up too_** on the page in the web browser.

This is what I mean about the difference in layout should reflect. For example, let's use `/code-collabo` routes. So on any page within code-collabo routes, We have sidebar (which houses the logo, and nav menu) and there is an header with h2 element that displays the page title dynamically. We want to have all these displayed together with the 404 dummy content specific to `/code-collabo` routes that says: "Code Collabo Subcommunity 404 page content".

You can use this same idea for the other layouts and 404 content.
`AppPageStructure.tsx` files should show you what I just described. Additionally you can check out the `appInfo.ts` files to see how pages, routes and page titles are exported.

samuko-things commented:
> **_Review comment 1 of 2_**
> 
> So this is a great start @samuko-things but **_the 404 pages don't work yet_**. **_How to test what I mean?_** - Before you created this 3 404.tsx files, there is a 404.tsx that is directly in the root of the pages folder. Remove that particular 404.tsx file and then try to access a page that does not exist in the browser. You will see that nextjs default way of showing 404 takes over.
> 
> That is to tell you that what is still showing up before was only that 404.tsx file that you deleted from the root of the pages folder.
> 
> Another confirmation is to restore the deleted 404.tsx file that is in the root of the pages folder. Check a non-exisiting route, you will see that only one text always shows up for 404: "Collabo Community Main App 404 page content". If you **_check the different 404 components you imported, you will see that different content is supposed to appear_** for the different types of 404.
> 
> **_Review comment 2 of 2_**
> 
> Also **_the difference in layout is supposed to reflect in the 404 pages too_** based on the part of the app that they take care of, then **_the specific dummy 404 page content should show up too_** on the page in the web browser.
> 
> This is what I mean about the difference in layout should reflect. For example, let's use `/code-collabo` routes. So on any page within code-collabo routes, We have sidebar (which houses the logo, and nav menu) and there is an header with h2 element that displays the page title dynamically. We want to have all these displayed together with the 404 dummy content specific to `/code-collabo` routes that says: "Code Collabo Subcommunity 404 page content".
> 
> You can use this same idea for the other layouts and 404 content. `AppPageStructure.tsx` files should show you what I just described. Additionally you can check out the `appInfo.ts` files to see how pages, routes and page titles are exported.

I've gone through your comment and tried somethings out, as well as researched somethings also about displaying custom404 pages in nextjs. 

the problem I think is that, nextjs only allows for one 404 page file, found in the root page folder, for static routes (e.g of dynamic route according to nextjs is ".../code-collabo/[:id]". any route that does not take a param is static route).

the solution I wanted to try was to be able to get the route entered from the browser and check the different route is an if statement and output the one with the corresponding route but I couldn't (I even messed up the whole branch that had to delete it and revert back to the current state).
I hope you understand, what I'm saying. I you have any suggested solution, do help me out.


** regarding comment two**
I think I understand what to do. I should send a PR soon. but without the implementation of the displaying different custom404
for now as I have not gotten a way round that.

Ifycode commented:
Yes, nextjs takes care of 404 through that one file. That's why I said to delete and restore to see the effect or outcome.

I also pointed you to some files so that it may give you inspiration of what may likely work.

Hoping to see the new changes.
Although I think you need to send a new PR, you can just update this one. Or is there a reason why you want to send a new PR? @samuko-things

Ifycode commented:
Yes, nextjs takes care of 404 through that one file. That's why I said to delete and restore to see the effect or outcome.

I also pointed you to some files so that it may give you inspiration of what may likely work.

Hoping to see the new changes.
Although I don't think you need to send a new PR, you can just update this one. Or is there a reason why you want to send a new PR? @samuko-things

samuko-things commented:
> Yes, nextjs takes care of 404 through that one file. That's why I said to delete and restore to see the effect or outcome.
> 
> I also pointed you to some files so that it may give you inspiration of what may likely work.
> 
> Hoping to see the new changes. Although I think you need to send a new PR, you can just update this one. Or is there a reason why you want to send a new PR? @samuko-things

no, I'm not sending a new PR, I'll update this PR

samuko-things committed:
update 404 file with basic styles

samuko-things commented:
@Ifycode, @code-collabo/community-mentors, I have just updated the 404 files for each route with basic styles as suggested in **Review comment 2 of 2**. pls do check it out

samuko-things commented:
I haven't implemented the **Review comment 1 of 2** yet.

samuko-things committed:
404-test-branch

Ifycode committed:
Complete 404 layout from where Samuel stopped

Ifycode reviewed (commented):
@samuko-things I have fixed the 404 layout and it's now working properly. I have merged it into this PR. So, you will need to pull in changes that I've merged into this `404-display` branch unto your local computer.

I'll be leaving you review comments to clean up the code and PR and any other thing needed.

Ifycode commented on client/src/apps/code-collabo/components/404component.tsx:
```diff
@@ -1,8 +1,12 @@
 
 export default function Custom404Component() {
+  let { pathname } = useRouter();
+  console.log(pathname);
```
1. You can safely remove this `useRouter` and `pathname` @samuko-things

Ifycode commented on client/src/apps/code-collabo/components/AppPageStructure.tsx:
```diff
@@ -14,6 +14,7 @@ export default function PageStructure({ children }: { children: ReactNode }) {
   let { pathname } = useRouter();
   pathname === urlStart ? pathname =  `${urlStart}/overview` : pathname;
   let { thisPage, pageTitle } = getPage(pathname, urlStart, appInfo.name);
+  console.log(appInfo)
```
2. Remove this `console.log` @samuko-things

Ifycode commented on client/src/apps/code-collabo/helpers/appInfo.ts:
```diff
@@ -7,6 +7,7 @@ const urlStart = stringToUrlStart(appName);
 const page = {
   1: 'overview',
   2: 'projects',
+  3: '404'
```
3. You can remove the `404` from among the pages @samuko-things

Ifycode commented on client/src/apps/code-collabo/helpers/appInfo.ts:
```diff
@@ -20,6 +21,10 @@ const appInfo: any = {
+    3: {
+      name: page[3],
+      route: `${urlStart}/${page[3]}`,
     }
```
4. You can remove the `404` object here too @samuko-things

Ifycode commented on client/src/apps/community/helpers/appInfo.ts:
```diff
@@ -16,6 +17,10 @@ const appInfo: any = {
+    2: {
+      name: page[2],
+      route: `${urlStart}/${page[2]}`,
+    },
```
5. You can remove anything `404` from this file too @samuko-things

Ifycode commented on client/src/apps/other-subcommunity/helpers/appInfo.ts:
```diff
@@ -16,6 +17,10 @@ const appInfo: any = {
+    2: {
+      name: page[2],
+      route: `${urlStart}/${page[2]}`,
+    },
```
6. You can remove anything `404` from this file too @samuko-things

Ifycode commented on client/src/pages/404-copy.tsx:
```diff
@@ -0,0 +1,24 @@
+
+
+OverviewPageCodeCollabo.getLayout = getCodeCollaboSubCommunityLayout;
+export default OverviewPageCodeCollabo;
```
7. You can safely remove remove this 404-copy.tsx file now @samuko-things

Ifycode commented on client/src/apps/code-collabo/components/404component.tsx:
```diff
@@ -1,8 +1,12 @@
   return (
     <>
-      <h1>Code Collabo Subcommunity 404 page content</h1>
+      <h1 className={`${custom404.test} dummy-module-text`}>Code Collabo Subcommunity 404 page content</h1>
```
8. For the 404 module styling, it's confusing to let all module.css bear the same name.
- module.css styles for community: Instead of `custom404.test` it can be `main404.test`
- module.css styles for code-collabo: Instead of `custom404.test` it can be `firstApp404.test`
- module.css styles for other-subcommunity: Instead of `custom404.test` it can be `secondApp404.test`

> Don't forget the comments about what it's imported as in the styles files too where necessary. Update the names everywhere as needed.

samuko-things committed:
done required cleanup

samuko-things commented:
@Ifycode, I've done the necessary cleanups like you asked. 
I've also gone through the updated solution. It's nice and brilliant.

samuko-things commented:
do let me know if I missed anything.

Ifycode commented:
Ok. Looks good @samuko-things 
Well done, good to know you find the solution nice.

One more thing. You will notice there are now conflicts with the develop branch. Pull the latest changes in develop and merge develop into this branch. Help resolve the conflicts, and test that the app still works appropriately afterwards.

samuko-things committed:
resolved conflit

samuko-things committed:
resolved confict

samuko-things commented:
> Ok. Looks good @samuko-things Well done, good to know you find the solution nice.
> 
> One more thing. You will notice there are now conflicts with the develop branch. Pull the latest changes in develop and merge develop into this branch. Help resolve the conflicts, and test that the app still works appropriately afterwards.

I've resolved the conflicts @Ifycode and also tested the application. everything still works as before.

Ifycode reviewed (approved):
Nice work @samuko-things
